### ISSUE TITLE
pip install after vaping results in munge 1.1 conflict

### ISSUE THREAD

**Issue by FliesLikeABrick** [20c/graphsrv#36]
This was with python 3.6.8 but may occur with later versions.

vaping 1.4.3 requires munge[tomlkit,yaml]<2.0.0,>=1.2.0, but you have munge 1.1.0 which is incompatible.

Steps to reproduce:
```
python3 -m venv vaping-dependencies
source vaping-dependences/bin/activate
python3 -m pip install --upgrade pip
python3 -m pip install vaping
vaping --home=vaping/ --debug
# errors about not having graphsrv, so we ran that as a separate pip install
python3 -m pip install graphsrv
# vaping 1.4.3 requires munge[tomlkit,yaml]<2.0.0,>=1.2.0, but you have munge 1.1.0 which is incompatible.
```

I haven't figured out which dependency in the original vaping install leads to 1.1.0 being installed from pypi instead of 1.2.0
this order of operations is what we ran into since we installed vaping, tried to run it with a config we had on hand, went "oh right we need graphsrv" when that failed to load -- and then ran into this situation

if we specify vaping and graphsrv to the same pip command, that leads to the extended pip runtime where it tries to find versions of a few packages that satisfy  overlapping subsets of version constraints (click, munge, jija2)

**Comment by FliesLikeABrick** [comment]
Full `pip install` output:
<details>
  <summary>Click to expand pip install output</summary>

(vaping-dependencies2) [user@mon101 site-packages]$ python3 -m pip install vaping graphsrv
Collecting vaping
  Using cached vaping-1.4.3-py3-none-any.whl (31 kB)
Collecting graphsrv
  Using cached graphsrv-2.0.0.tar.gz (194 kB)
Collecting python-daemon<3,>=2
  Using cached python_daemon-2.3.0-py2.py3-none-any.whl (35 kB)
Requirement already satisfied: munge[tomlkit,yaml]<2.0.0,>=1.2.0 in ./vaping-dependencies2/lib/python3.6/site-packages (from vaping) (1.2.0)
Collecting confu<2.0.0,>=1.7.1
  Using cached confu-1.8.0-py3-none-any.whl (18 kB)
Collecting pid<4,>=3
  Using cached pid-3.0.4-py2.py3-none-any.whl (11 kB)
Requirement already satisfied: pluginmgr<2,>=1 in ./vaping-dependencies2/lib/python3.6/site-packages (from vaping) (1.0.1)
Requirement already satisfied: vodka<4,>=3 in ./vaping-dependencies2/lib/python3.6/site-packages (from graphsrv) (3.1.0)
Requirement already satisfied: Jinja2<3.0,>=2.8 in ./vaping-dependencies2/lib/python3.6/site-packages (from graphsrv) (2.11.3)
Collecting Flask<2,>=0.11.1
  Using cached Flask-1.1.4-py2.py3-none-any.whl (94 kB)
Collecting click<8.0,>=5.1
  Using cached click-7.1.2-py2.py3-none-any.whl (82 kB)
Collecting itsdangerous<2.0,>=0.24
  Using cached itsdangerous-1.1.0-py2.py3-none-any.whl (16 kB)
Collecting Werkzeug<2.0,>=0.15
  Using cached Werkzeug-1.0.1-py2.py3-none-any.whl (298 kB)
Requirement already satisfied: MarkupSafe>=0.23 in ./vaping-dependencies2/lib/python3.6/site-packages (from Jinja2<3.0,>=2.8->graphsrv) (2.0.1)
Requirement already satisfied: requests<3.0,>=2.6 in ./vaping-dependencies2/lib/python3.6/site-packages (from munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (2.26.0)
INFO: pip is looking at multiple versions of markupsafe to determine which version is compatible with other requirements. This could take a while.
Collecting MarkupSafe>=0.23
  Using cached MarkupSafe-2.0.1-cp36-cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (30 kB)
  Using cached MarkupSafe-2.0.0-cp36-cp36m-manylinux2010_x86_64.whl (30 kB)
  Using cached MarkupSafe-1.1.1-cp36-cp36m-manylinux2010_x86_64.whl (32 kB)
  Using cached MarkupSafe-1.1.0-cp36-cp36m-manylinux1_x86_64.whl (27 kB)
  Using cached MarkupSafe-1.0.tar.gz (14 kB)
  Using cached MarkupSafe-0.23.tar.gz (13 kB)
INFO: pip is looking at multiple versions of jinja2 to determine which version is compatible with other requirements. This could take a while.
Collecting Jinja2<3.0,>=2.8
  Using cached Jinja2-2.11.3-py2.py3-none-any.whl (125 kB)
INFO: pip is looking at multiple versions of markupsafe to determine which version is compatible with other requirements. This could take a while.
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
  Using cached Jinja2-2.11.2-py2.py3-none-any.whl (125 kB)
  Using cached Jinja2-2.11.1-py2.py3-none-any.whl (126 kB)
  Using cached Jinja2-2.11.0-py2.py3-none-any.whl (126 kB)
  Using cached Jinja2-2.10.3-py2.py3-none-any.whl (125 kB)
  Using cached Jinja2-2.10.2-py2.py3-none-any.whl (125 kB)
  Using cached Jinja2-2.10.1-py2.py3-none-any.whl (124 kB)
INFO: pip is looking at multiple versions of jinja2 to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of itsdangerous to determine which version is compatible with other requirements. This could take a while.
Collecting itsdangerous<2.0,>=0.24
  Using cached itsdangerous-0.24.tar.gz (46 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: pip is looking at multiple versions of click to determine which version is compatible with other requirements. This could take a while.
Collecting click<8.0,>=5.1
  Using cached click-7.1.1-py2.py3-none-any.whl (82 kB)
  Using cached click-7.1-py2.py3-none-any.whl (82 kB)
  Using cached Click-7.0-py2.py3-none-any.whl (81 kB)
INFO: pip is looking at multiple versions of itsdangerous to determine which version is compatible with other requirements. This could take a while.
  Using cached click-6.7-py2.py3-none-any.whl (71 kB)
  Using cached click-6.6-py2.py3-none-any.whl (71 kB)
  Using cached click-6.5.tar.gz (283 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
  Using cached click-6.4-py2.py3-none-any.whl (70 kB)
INFO: pip is looking at multiple versions of click to determine which version is compatible with other requirements. This could take a while.
  Using cached click-6.3-py2.py3-none-any.whl (70 kB)
  Using cached click-6.2-py2.py3-none-any.whl (70 kB)
  Using cached click-6.1-py2.py3-none-any.whl (70 kB)
  Using cached click-6.0-py2.py3-none-any.whl (70 kB)
  Using cached click-5.1-py2.py3-none-any.whl (65 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: pip is looking at multiple versions of flask to determine which version is compatible with other requirements. This could take a while.
Collecting Flask<2,>=0.11.1
  Using cached Flask-1.1.3-py2.py3-none-any.whl (94 kB)
  Using cached Flask-1.1.2-py2.py3-none-any.whl (94 kB)
Requirement already satisfied: click>=5.1 in ./vaping-dependencies2/lib/python3.6/site-packages (from Flask<2,>=0.11.1->graphsrv) (8.0.1)
Requirement already satisfied: Werkzeug>=0.15 in ./vaping-dependencies2/lib/python3.6/site-packages (from Flask<2,>=0.11.1->graphsrv) (2.0.1)
Requirement already satisfied: itsdangerous>=0.24 in ./vaping-dependencies2/lib/python3.6/site-packages (from Flask<2,>=0.11.1->graphsrv) (2.0.1)
Requirement already satisfied: importlib-metadata in ./vaping-dependencies2/lib/python3.6/site-packages (from click>=5.1->Flask<2,>=0.11.1->graphsrv) (4.6.4)
Requirement already satisfied: PyYAML<6.0,>=5.1 in ./vaping-dependencies2/lib/python3.6/site-packages (from munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (5.4.1)
Requirement already satisfied: tomlkit<0.8.0,>=0.7.2 in ./vaping-dependencies2/lib/python3.6/site-packages (from munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (0.7.2)
Requirement already satisfied: docutils in ./vaping-dependencies2/lib/python3.6/site-packages (from python-daemon<3,>=2->vaping) (0.17.1)
Requirement already satisfied: setuptools in ./vaping-dependencies2/lib/python3.6/site-packages (from python-daemon<3,>=2->vaping) (36.5.0)
Requirement already satisfied: lockfile>=0.10 in ./vaping-dependencies2/lib/python3.6/site-packages (from python-daemon<3,>=2->vaping) (0.12.2)
Requirement already satisfied: certifi>=2017.4.17 in ./vaping-dependencies2/lib/python3.6/site-packages (from requests<3.0,>=2.6->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (2021.5.30)
Requirement already satisfied: charset-normalizer~=2.0.0 in ./vaping-dependencies2/lib/python3.6/site-packages (from requests<3.0,>=2.6->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (2.0.4)
Requirement already satisfied: idna<4,>=2.5 in ./vaping-dependencies2/lib/python3.6/site-packages (from requests<3.0,>=2.6->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (3.2)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in ./vaping-dependencies2/lib/python3.6/site-packages (from requests<3.0,>=2.6->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (1.26.6)
Requirement already satisfied: tmpl<2,>=1 in ./vaping-dependencies2/lib/python3.6/site-packages (from vodka<4,>=3->graphsrv) (1.0.0)
Requirement already satisfied: dataclasses in ./vaping-dependencies2/lib/python3.6/site-packages (from Werkzeug>=0.15->Flask<2,>=0.11.1->graphsrv) (0.8)
Requirement already satisfied: zipp>=0.5 in ./vaping-dependencies2/lib/python3.6/site-packages (from importlib-metadata->click>=5.1->Flask<2,>=0.11.1->graphsrv) (3.5.0)
Requirement already satisfied: typing-extensions>=3.6.4 in ./vaping-dependencies2/lib/python3.6/site-packages (from importlib-metadata->click>=5.1->Flask<2,>=0.11.1->graphsrv) (3.10.0.0)
Using legacy 'setup.py install' for graphsrv, since package 'wheel' is not installed.
Installing collected packages: python-daemon, pid, Flask, confu, vaping, graphsrv
    Running setup.py install for graphsrv ... done
Successfully installed Flask-1.1.2 confu-1.8.0 graphsrv-2.0.0 pid-3.0.4 python-daemon-2.3.0 vaping-1.4.3
(vaping-dependencies2) [user@mon101 site-packages]$ 
</details>

**Comment by vegu** [comment]
Thanks for the report, will look into asap.

**Comment by vegu** [comment]
Looks it reverts back to munge 1.1 because munge 1.2 needs click >= 8, while the flask version pinned by graphsrv (flask 1.1.4) requires click<8

**Comment by FliesLikeABrick** [comment]
That explains why, when you specify "pip install vaping graphsrv" -- pip spends a while trying to resolve dependencies, including click and munge (trying to find versions of each package that have compatible version pinning).  

Possible resolutions...
- can graphsrv use a newer flask that uses click >= 8?
- Can munge 1.2 use click < 8 ?

**Comment by vegu** [comment]
We are testing graphsrv with Flask 2, change is available via the `flask-upgrade` branch while we check for potential issues.

```
pip install https://github.com/20c/graphsrv/archive/refs/heads/flask-upgrade.zip
```

**Comment by vegu** [comment]
We adjusted the dependency in munge and released munge 1.2.1 which should fix the issue

Will leave this ticket open since moving graphsrv to Flask2 seems like a good idea regardless (and seems to be functional)

**Comment by FliesLikeABrick** [comment]
Confirmed fixed in pypi packages, and indeed graphsrv still downgrades click to 7.1.2 due to Flask 1 being used
